async-stream-impl/src/lib.rs

```diff
@@ -120,10 +120,19 @@ impl VisitMut for Scrub<'_> {
 
                 // let ident = &self.yielder;
 
-                *i = if self.is_try {
-                    syn::parse_quote! { __yield_tx.send(::core::result::Result::Ok(#value_expr)).await }
+                let yield_expr = if self.is_try {
+                    quote! { __yield_tx.send(::core::result::Result::Ok(#value_expr)).await }
                 } else {
-                    syn::parse_quote! { __yield_tx.send(#value_expr).await }
+                    quote! { __yield_tx.send(#value_expr).await }
+                };
+                *i = syn::parse_quote! {
+                    {
+                        #[allow(unreachable_code)]
+                        if false {
+                            break 'check_stream_scope (loop {});
+                        }
```

Comment:
Perhaps this can still be bypassed by defining a block labeled `'check_stream_scope` in the async block.

That said, if we rename the label to something that clearly indicates that it is a async-stream's private API (something like `'__async_stream_private_check_scope`), that would actually be sufficient, since only “users who explicitly abuse private APIs” would be affected.